index.html

```diff
@@ -159,141 +156,401 @@ <h2>
+                    </li>
+                    <li>[=Reject=] |promise| with a
+                    {{"WrongDocumentError"}} {{DOMException}}.
+                    </li>
+                  </ol>
+                </li>
+                <li>Return <var>promise</var>.
+                </li>
+              </ol>
+            </li>
+            <li>If the user agent's [=pointer-lock target=] is an element whose
+            [=shadow-including root=] is not equal to [=this=]'s
+            [=shadow-including root=], then:
+              <ol>
+                <li>[=Queue an element task=] on the [=user interaction task
+                source=], given this, to perform the following steps:
+                  <ol>
+                    <li>[=Fire an event=] named {{pointerlockerror}} at
+                    [=this=]'s [=Node/node document=].
+                    </li>
+                    <li>[=Reject=] <var>promise</var> with a
+                    "{{InvalidStateError}}" {{DOMException}}.
+                    </li>
+                  </ol>
+                </li>
+                <li>Return <var>promise</var>.
+                </li>
+              </ol>
+            </li>
+            <li>If the user agent's [=pointer-lock target=]'s
+            [=shadow-including root=] is equal to [=this=]'s [=shadow-including
+            root=], then:
+              <ol>
+                <li>If [=options=] are equivalent to the current [=pointer-lock
+                options=]
+                  <ol>
+                    <li>Set [=pointer-lock target=] to [=this=]
+                    </li>
+                    <li>[=Queue an element task=] on the [=user interaction
+                    task source=], given this, to perform the following steps:
+                      <ol>
+                        <li>[=Fire an event=] named {{pointerlockchange}} at
+                        [=this=]'s [=Node/node document=].
+                        </li>
+                        <li>[=Resolve=] the <var>promise</var>.
+                        </li>
+                      </ol>
+                    </li>
+                  </ol>
+                </li>
+                <li>If [=options=] are not equivalent to the user agent's
+                current [=pointer-lock options=] and [=options=] are not
+                supported:
+                  <ol>
+                    <li>[=Queue an element task=] on the [=user interaction
+                    task source=], given this, to perform the following steps:
+                      <ol type='i'>
+                        <li>[=Reject=] <var>promise</var> with a
+                        "{{NotSupportedError}}" {{DOMException}}.
+                        </li>
+                      </ol>
+                    </li>
+                    <li>Return <var>promise</var>.
+                    </li>
+                  </ol>
+                </li>
+                <li>If [=options=] are not equivalent to the current
```

Comment:
We don't think we need to go here is we prevent the method from being called more than once. 

Having a flag will prevent getting into this weird override with supported situation. 

Let's please simplify all this unless there is an actual use case for calling lock more than once without unlocking first.

Reply:
Are you asking why an application would want to change their unadjustedMovement flag without unlocking and re-locking?

As in, is there is a use case for the following pattern?

```
element.requestPointerLock({unadjustedMovement: false})
// ... some time later, but still with the pointer locked
element.requestPointerLock({unadjustedMovement: true})
// ... some more time later, but still with the pointer locked
element.requestPointerLock({unadjustedMovement: false})
```

We use this pattern in cloud gaming. The reason is even when the mouse cursor is visible (you're using the mouse as a pointer), we need to confine the cursor within the window for certain types of games (e.g. real time strategy games that scroll the map when the cursor is placed at the edge of the window). So in that case we want accelerated movement to match mouse feel at the desktop, but we move a cursor image around the screen rather than relinquishing pointer lock. When the game requests the cursor be hidden and locked such as for 1st person 3D camera control, we switch the unadjustedMovement flag to true so that the 3D camera control feels natural. When the game exits this mode and unhides the game cursor, we go back to unadjustedMovement = false so that the cursor control feels right for use as a pointer again, but we continue with pointer lock so the cursor cannot exit the browser window.

If we didn't have the ability to change unadjustedMovement without relinquishing pointer lock, we would run the risk that in between unlocking and re-locking the cursor that the cursor could escape the window and the user would inadvertently click on something else. There are games that rapidly swap between unaccelerated (for 3D camera control) and accelerated (for pointer control) that can get intense and losing focus due to the cursor escaping the window would be a big problem.

I can think of some other solutions to these problems, but they would be quite large changes to the spec. 

One potential solution would be to not have an unadjustedMovement option at all, but instead just include unacceleratedMovementX/unacceleratedMovementY on all events that user agents could populate. One reason I suspect that wasn't done in the first place is on some platforms (e.g. Windows), it's not easy to correlate accelerated movement events with unaccelerated movement events (they come from different subsystems at possibly different times).

Another potential solution would be to add API for mouse confinement, which the spec [currently notes as a possible future enhancement](https://www.w3.org/TR/pointerlock-2/#why-bundle), that could then be used in lieu of unacceleratedMovement = false in the application described above. This would actually be best, because the difference between the native hardware cursor vs a cursor image being repositioned by javascript is something you can feel, but of course it would be a big change and it's somewhat of a separate topic.

---

P.S Sorry if I just wrote a novel in response to the wrong question, but I wanted to jump in here and represent our use for changing movement types without relinquishing pointer lock rather than lose it in the final version of the spec.

Reply:
Thanks for the clarification! that's really helpful and make sense for when the promises have resolved and switching is required (i.e., "happy path uses cases"). 

I'm was just thinking about really simple sync cases (i.e., someone forgets to `await` or consider the following):

```JS
const p1 = element.requestPointerLock({unadjustedMovement: false});
const p2  = element.requestPointerLock({unadjustedMovement: true});
const p3 = element.requestPointerLock({unadjustedMovement: false});

// What happens to p1, p2, and p3? 
```

in the most simple cases, p2 and p3 should be rejected because `p1` hasn't settled yet.

Reply:
In my view, p1, p2, and p3 all resolve in order (assuming all other requirements for success are met). I think of the requests as queuing up. The fact that the first two are replaced by the third and aren't active long enough to have any mouse events associated with them isn't an error itself.

If unadjustedMovement isn't supported, then I would expect p1 to resolve, p2 to reject (with NotSupportedError), and then p3 to resolve. My reasoning here is that when options aren't supported, the only action to take is to reject the promise, not to unlock or change the target element.

If there is some other error condition, like we don't have user interaction or the window isn't in focus, then all 3 should reject with that reason.

I think this interpretation is consistent with what's been written so far in this PR.

Reply:
Use cases served:

1. Changing options.
2. Changing lock targets (including across shadow dom)

I think https://en.wikipedia.org/wiki/Robustness_principle applies here.

Current wording has evolved substantially from the [original](https://www.w3.org/TR/pointerlock/#extensions-to-the-element-interface).

> If any element (including this one), whose shadow-including root is same as this element's shadow-including root, is already locked (or pending lock) the pointer lock target must be updated to this element and a pointerlockchange event sent.


Pre-promise there was not a concern of having multiple asynchronous requests, as they would only resolve eventually with an event.
```js
// Assume an unlocked state, and that these lines execute before the lock is granted.
element1.requestPointerLock(options);
element2.requestPointerLock(options);
// expect change event with element2 as target.
```

Now with promises this is more subtle: 
```js
// Assume an unlocked state, and that these lines execute before the lock is granted.
const promise1 = element1.requestPointerLock(options);
const promise2 = element2.requestPointerLock(options);
```

Mostly, a consistent result needs to be specified. Ideally consistent with current era promise design.

I'd personally (acknowledging that I'm not active in many current specs) suggest:

All promises resolve truthy. There was no reason to block the request. Any code on page expecting success shouldn't be told it failed, because overall lock was obtained.

Before promises the code would have been interpreted as successful. Adding promises shouldn't change that.  We should avoid conflicting error/success state between events and promises.

Reply:
I agree with Vincent that maintaining support for both pre- and post-promise versions of `requestPointerLock` is essential. Instead of rejecting concurrent requests, we should implement a requests queue and handle them sequentially.

Reply:
Will re-review once the above is integrated. We agreed queuing is fine.